{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hidden Markov Models"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "author: Jacob Schreiber <br>\n",
    "contact: jmschreiber91@gmail.com"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hidden Markov models (HMMs) are the flagship of the pomegranate package, in that most time is spent improving their implementation, and these improvements sometimes trickle down into the other algorithms. Lets delve into the features which pomegranate offers.\n",
    "\n",
    "Hidden Markov models are a form of structured prediction method which extend general mixture models to sequences of data, where position in the sequence is relevant. If each point in this sequence is completely independent of the other points, then HMMs are not the right tools and GMMs (or more complicated Bayesian networks) may be a better tool.\n",
    "\n",
    "The most common examples of HMMs come from bioinformatics and natural language processing. Since I am a bioinformatician, I will predominately use examples from bioinformatics.\n",
    "\n",
    "## GMMs vs HMMs in the CG rich region example\n",
    "\n",
    "Lets take the simplified example of CG island detection on a sequence of DNA. DNA is made up of the four canonical nucleotides, abbreviated 'A', 'C', 'G', and 'T'. Specific organizations of these nucleotides encode enough information to build you, a human being. One simple region in the genome is called the 'CG' island, where the nucleotides 'C' and 'G' are enriched. Lets compare the predictions of a GMM with the predictions of a HMM, to both understand conceptually the differences between the two, and to see how easy it is to use pomegranate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: pylab import has clobbered these variables: ['log', 'random']\n",
      "`%matplotlib` prevents importing * from pylab and numpy\n"
     ]
    }
   ],
   "source": [
    "from pomegranate import *\n",
    "import numpy as np\n",
    "%pylab inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "seq = list('CGACTACTGACTACTCGCCGACGCGACTGCCGTCTATACTGCGCATACGGC')\n",
    "\n",
    "d1 = DiscreteDistribution({'A': 0.25, 'C': 0.25, 'G': 0.25, 'T': 0.25})\n",
    "d2 = DiscreteDistribution({'A': 0.10, 'C': 0.40, 'G': 0.40, 'T': 0.10})\n",
    "\n",
    "s1 = State( d1, name='background' )\n",
    "s2 = State( d2, name='CG island' )\n",
    "\n",
    "gmm = GeneralMixtureModel( [d1, d2] )\n",
    "hmm = HiddenMarkovModel()\n",
    "hmm.add_states(s1, s2)\n",
    "hmm.add_transition( hmm.start, s1, 0.5 )\n",
    "hmm.add_transition( hmm.start, s2, 0.5 )\n",
    "hmm.add_transition( s1, s1, 0.5 )\n",
    "hmm.add_transition( s1, s2, 0.5 )\n",
    "hmm.add_transition( s2, s1, 0.5 )\n",
    "hmm.add_transition( s2, s2, 0.5 )\n",
    "hmm.bake()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence: CGACTACTGACTACTCGCCGACGCGACTGCCGTCTATACTGCGCATACGGC\n",
      "gmm pred: 110100101010010111110111101011110100001011110001111\n",
      "hmm pred: 110100101010010111110111101011110100001011110001111\n"
     ]
    }
   ],
   "source": [
    "gmm_predictions = gmm.predict( np.array(seq) )\n",
    "hmm_predictions = hmm.predict( seq )\n",
    "\n",
    "print \"sequence: {}\".format( ''.join( seq ) )\n",
    "print \"gmm pred: {}\".format( ''.join( map( str, gmm_predictions ) ) )\n",
    "print \"hmm pred: {}\".format( ''.join( map( str, hmm_predictions ) ) )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note: The HMM and GMM predictions may be the inverse of each other, because HMM states undergo a topological sort in order to properly handle silent states (more later), which can cause the order they were inserted into the model. \n",
    "\n",
    "Your first reaction may to say \"But Jacob, you just said that HMMs and GMMs are different. Why shold I make a HMM when making a GMM is so easy?\". \n",
    "\n",
    "My point in showing you this is that a dense HMM with equal probabilities between each state is ~equivalent~ to a GMM. However, this framework gives us great flexibility to add prior knowledge, whereas a GMM doesn't. If we look at the predictions, we see that it's bifurcating between \"background\" and \"CG island\" very quickly--in essence, calling every C or G a 'CG island'. This is not likely to be true. We know that CG islands have some As and Ts in them, and background sequence has Cs and Gs. We can change the transition probabilities to account for this, and prevent switching from occuring too rapidly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "hmm = HiddenMarkovModel()\n",
    "hmm.add_states(s1, s2)\n",
    "hmm.add_transition( hmm.start, s1, 0.5 )\n",
    "hmm.add_transition( hmm.start, s2, 0.5 )\n",
    "hmm.add_transition( s1, s1, 0.9 )\n",
    "hmm.add_transition( s1, s2, 0.1 )\n",
    "hmm.add_transition( s2, s1, 0.1 )\n",
    "hmm.add_transition( s2, s2, 0.9 )\n",
    "hmm.bake()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence: CGACTACTGACTACTCGCCGACGCGACTGCCGTCTATACTGCGCATACGGC\n",
      "gmm pred: 110100101010010111110111101011110100001011110001111\n",
      "hmm pred: 000000000000000111111111111111100000000000000001111\n",
      "\n",
      "hmm state 0: background\n",
      "hmm state 1: CG island\n"
     ]
    }
   ],
   "source": [
    "hmm_predictions = hmm.predict( seq )\n",
    "\n",
    "print \"sequence: {}\".format( ''.join( seq ) )\n",
    "print \"gmm pred: {}\".format( ''.join( map( str, gmm_predictions ) ) )\n",
    "print \"hmm pred: {}\".format( ''.join( map( str, hmm_predictions ) ) )\n",
    "print\n",
    "print \"hmm state 0: {}\".format( hmm.states[0].name )\n",
    "print \"hmm state 1: {}\".format( hmm.states[1].name )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This seems far more reasonable. There is a single CG island surrounded by background sequence, and something at the end. If we knew that CG islands cannot occur at the end of sequences, we need only modify the underlying structure of the HMM in order to say that the sequence must end from the background state."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "hmm = HiddenMarkovModel()\n",
    "hmm.add_states(s1, s2)\n",
    "hmm.add_transition( hmm.start, s1, 0.5 )\n",
    "hmm.add_transition( hmm.start, s2, 0.5 )\n",
    "hmm.add_transition( s1, s1, 0.89 )\n",
    "hmm.add_transition( s1, s2, 0.10 )\n",
    "hmm.add_transition( s1, hmm.end, 0.01 )\n",
    "hmm.add_transition( s2, s1, 0.1 )\n",
    "hmm.add_transition( s2, s2, 0.9 )\n",
    "hmm.bake()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence: CGACTACTGACTACTCGCCGACGCGACTGCCGTCTATACTGCGCATACGGC\n",
      "gmm pred: 110100101010010111110111101011110100001011110001111\n",
      "hmm pred: 000000000000000111111111111111100000000000000000000\n",
      "\n",
      "hmm state 0: background\n",
      "hmm state 1: CG island\n"
     ]
    }
   ],
   "source": [
    "hmm_predictions = hmm.predict( seq )\n",
    "\n",
    "print \"sequence: {}\".format( ''.join( seq ) )\n",
    "print \"gmm pred: {}\".format( ''.join( map( str, gmm_predictions ) ) )\n",
    "print \"hmm pred: {}\".format( ''.join( map( str, hmm_predictions ) ) )\n",
    "print\n",
    "print \"hmm state 0: {}\".format( hmm.states[0].name )\n",
    "print \"hmm state 1: {}\".format( hmm.states[1].name )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looks like we managed to get rid of that pesky end (again, the numbers may have flipped, look at the indices). Modifying transition probabilities and using non-dense graphical structures are two major ways in which HMMs account for data in a sequence not being independent and identically distributed (i.i.d.). In fact, in most applications, the graphical structure of a HMM is very sparse. \n",
    "\n",
    "If we want a more probabilistic view of what's going on, we can get the probability of each symbol in the sequence being in each of the states in the model easily. This is useful to get a soft estimate of classification, which allows us to include confidence as well as prediction. Values close to 50-50 get masked when you make hard classifications, but this uncertainty can be passed to future applications if you use soft assignments. Each row in the matrix is one symbol in the sequence, and the columns correspond to the two states identified above (CG island or background)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 0.5172946   0.4827054 ]\n",
      " [ 0.57795936  0.42204064]\n",
      " [ 0.71401225  0.28598775]\n",
      " [ 0.74243544  0.25756456]\n",
      " [ 0.83959962  0.16040038]\n",
      " [ 0.86128017  0.13871983]\n",
      " [ 0.82782146  0.17217854]\n",
      " [ 0.85249835  0.14750165]\n",
      " [ 0.81978814  0.18021186]\n",
      " [ 0.84553819  0.15446181]\n",
      " [ 0.81211956  0.18788044]\n",
      " [ 0.83747113  0.16252887]\n",
      " [ 0.80158912  0.19841088]\n",
      " [ 0.67080299  0.32919701]\n",
      " [ 0.61633927  0.38366073]\n",
      " [ 0.41955381  0.58044619]\n",
      " [ 0.30924476  0.69075524]\n",
      " [ 0.25346817  0.74653183]\n",
      " [ 0.23607192  0.76392808]\n",
      " [ 0.2520183   0.7479817 ]\n",
      " [ 0.30592516  0.69407484]\n",
      " [ 0.25238171  0.74761829]\n",
      " [ 0.23678405  0.76321595]\n",
      " [ 0.25461533  0.74538467]\n",
      " [ 0.31103922  0.68896078]\n",
      " [ 0.42239529  0.57760471]\n",
      " [ 0.41608533  0.58391467]\n",
      " [ 0.46046222  0.53953778]\n",
      " [ 0.3969169   0.6030831 ]\n",
      " [ 0.38483311  0.61516689]\n",
      " [ 0.42071153  0.57928847]\n",
      " [ 0.51494207  0.48505793]\n",
      " [ 0.69481256  0.30518744]\n",
      " [ 0.74620572  0.25379428]\n",
      " [ 0.87389253  0.12610747]\n",
      " [ 0.91894035  0.08105965]\n",
      " [ 0.92362066  0.07637934]\n",
      " [ 0.89232532  0.10767468]\n",
      " [ 0.79568775  0.20431225]\n",
      " [ 0.76726124  0.23273876]\n",
      " [ 0.64148039  0.35851961]\n",
      " [ 0.59014274  0.40985726]\n",
      " [ 0.59838163  0.40161837]\n",
      " [ 0.66858294  0.33141706]\n",
      " [ 0.82107597  0.17892403]\n",
      " [ 0.87149451  0.12850549]\n",
      " [ 0.86714974  0.13285026]\n",
      " [ 0.80396469  0.19603531]\n",
      " [ 0.80239569  0.19760431]\n",
      " [ 0.86198839  0.13801161]\n",
      " [ 1.          0.        ]]\n"
     ]
    }
   ],
   "source": [
    "print hmm.predict_proba( seq )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There is a corresponding `hmm.predict_log_proba` method present if you want to get the log values. These are the emission probability values calculated by the forward backward algorithm, and can also be retrieved by calling `hmm.forward_backward( seq )`, which returns both the emission and the transition probability tables.\n",
    "\n",
    "Lets take a look at these tables!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 28.91051356   2.41288774   1.           0.        ]\n",
      " [  2.89559314  15.78100555   0.           0.        ]\n",
      " [  0.           0.           0.           0.        ]\n",
      " [  0.5172946    0.4827054    0.           0.        ]]\n"
     ]
    }
   ],
   "source": [
    "trans, ems = hmm.forward_backward( seq )\n",
    "print trans"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is the transition table, which has the soft count of the number of transitions across an edge in the model given a single sequence. It is a square matrix of size equal to the number of states (including start and end state), with number of transitions from (row_id) to (column_id). This is exemplified by the 1.0 in the first row, indicating that there is one transition from background state to the end state, as that's the only way to reach the end state. However, the third (or fourth, depending on ordering) row is the transitions from the start state, and it only slightly favors the background state. These counts are not normalized to the length of the input sequence, but can easily be done so by dividing by row sums, column sums, or entire table sums, depending on your application.\n",
    "\n",
    "A possible reason not to normalize is to run several sequences through and add up their tables, because normalizing in the end and extracting some domain knowledge. It is extremely useful in practice. For example, we can see that there is an expectation of 2.8956 transitions from CG island to background, and 2.4 from background to CG island. This could be used to infer that there are ~2-3 edges, which makes sense if you consider that the start and end of the sequence seem like they might be part of the CG island states except for the strict transition probabilities used (look at the first few rows of the emission table above.)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We've been using the forward backward algorithm and maximum a posteriori for decoding thus far, however maximum a posteriori decoding has the side effect that it is possible that it predicts impossible sequences in some edge cases. An alternative is Viterbi decoding, which at each step takes the most likely path, instead of sum of all paths, to produce hard assignments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence: CGACTACTGACTACTCGCCGACGCGACTGCCGTCTATACTGCGCATACGGC\n",
      "gmm pred: 110100101010010111110111101011110100001011110001111\n",
      "hmm pred: 000000000000000000000000000000000000000000000000000\n",
      "\n",
      "hmm state 0: background\n",
      "hmm state 1: CG island\n"
     ]
    }
   ],
   "source": [
    "hmm_predictions = hmm.predict( seq, algorithm='viterbi' )[1:-1]\n",
    "\n",
    "print \"sequence: {}\".format( ''.join( seq ) )\n",
    "print \"gmm pred: {}\".format( ''.join( map( str, gmm_predictions ) ) )\n",
    "print \"hmm pred: {}\".format( ''.join( map( str, hmm_predictions ) ) )\n",
    "print\n",
    "print \"hmm state 0: {}\".format( hmm.states[0].name )\n",
    "print \"hmm state 1: {}\".format( hmm.states[1].name )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see here a case in which it does not do too well. The Viterbi path can be more conservative in its transitions due to the hard assignments it makes. In essence, if multiple possibile paths are possible at a given point, it takes the most likely path, even if the sum of all other paths is greater than the sum of that path. In problems with a lower signal to noise ratio, this can mask the signal. As a side note, we can use the following to get the maximum a posteriori and Viterbi paths:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "v_logp, v_seq = hmm.viterbi( seq )\n",
    "m_logp, m_seq = hmm.maximum_a_posteriori( seq )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The sequence predicted is a tuple of (state id, state object) for every state in the predicted path. The predict method simply takes the state ids from this path and returns those as an array. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence Alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lets move on to a more complicated structure, that of a profile HMM. A profile HMM is used to align a sequence to a reference 'profile', where the reference profile can either be a single sequence, or an alignment of many sequences (such as a reference genome). In essence, this profile has a 'match' state for every position in the reference profile, and 'insert' state, and a 'delete' state. The insert state allows the external sequence to have an insertion into the sequence without throwing off the entire alignment, such as the following:\n",
    "\n",
    "`ACCG : Sequence` <br>\n",
    "`|| |` <br>\n",
    "`AC-G : Reference`\n",
    "\n",
    "or a deletion, which is the opposite:\n",
    "\n",
    "`A-G : Sequence` <br>\n",
    "`| |` <br>\n",
    "`ACG : Reference`\n",
    "\n",
    "The bars in the middle refer to a perfect match, whereas the lack of a bar means either a deletion/insertion, or a mismatch. A mismatch is where two positions are aligned together, but do not match. This models the biological phenomena of mutation, where one nucleotide can convert to another over time. It is usually more likely in biological sequences that this type of mutation occurs than that the nucleotide was deleted from the sequence (shifting all nucleotides over by one) and then another was inserted at the exact location (moving all nucleotides over again). Since we are using a probabilistic model, we get to define these probabilities through the use of distributions! If we want to model mismatches, we can just set our 'match' state to have an appropriate distribution with non-zero probabilities over mismatches. \n",
    "\n",
    "Lets now create a three nucleotide profile HMM, which models the sequence 'ACT'. We will fuzz this a little bit in the match states, pretending to have some prior information about what mutations occur at each position. If you don't have any information, setting a uniform, small, value over the other values is usually okay."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "model = HiddenMarkovModel( \"Global Alignment\")\n",
    "\n",
    "# Define the distribution for insertions\n",
    "i_d = DiscreteDistribution( { 'A': 0.25, 'C': 0.25, 'G': 0.25, 'T': 0.25 } )\n",
    "\n",
    "# Create the insert states\n",
    "i0 = State( i_d, name=\"I0\" )\n",
    "i1 = State( i_d, name=\"I1\" )\n",
    "i2 = State( i_d, name=\"I2\" )\n",
    "i3 = State( i_d, name=\"I3\" )\n",
    "\n",
    "# Create the match states\n",
    "m1 = State( DiscreteDistribution({ \"A\": 0.95, 'C': 0.01, 'G': 0.01, 'T': 0.02 }) , name=\"M1\" )\n",
    "m2 = State( DiscreteDistribution({ \"A\": 0.003, 'C': 0.99, 'G': 0.003, 'T': 0.004 }) , name=\"M2\" )\n",
    "m3 = State( DiscreteDistribution({ \"A\": 0.01, 'C': 0.01, 'G': 0.01, 'T': 0.97 }) , name=\"M3\" )\n",
    "\n",
    "# Create the delete states\n",
    "d1 = State( None, name=\"D1\" )\n",
    "d2 = State( None, name=\"D2\" )\n",
    "d3 = State( None, name=\"D3\" )\n",
    "\n",
    "# Add all the states to the model\n",
    "model.add_states( [i0, i1, i2, i3, m1, m2, m3, d1, d2, d3 ] )\n",
    "\n",
    "# Create transitions from match states\n",
    "model.add_transition( model.start, m1, 0.9 )\n",
    "model.add_transition( model.start, i0, 0.1 )\n",
    "model.add_transition( m1, m2, 0.9 )\n",
    "model.add_transition( m1, i1, 0.05 )\n",
    "model.add_transition( m1, d2, 0.05 )\n",
    "model.add_transition( m2, m3, 0.9 )\n",
    "model.add_transition( m2, i2, 0.05 )\n",
    "model.add_transition( m2, d3, 0.05 )\n",
    "model.add_transition( m3, model.end, 0.9 )\n",
    "model.add_transition( m3, i3, 0.1 )\n",
    "\n",
    "# Create transitions from insert states\n",
    "model.add_transition( i0, i0, 0.70 )\n",
    "model.add_transition( i0, d1, 0.15 )\n",
    "model.add_transition( i0, m1, 0.15 )\n",
    "\n",
    "model.add_transition( i1, i1, 0.70 )\n",
    "model.add_transition( i1, d2, 0.15 )\n",
    "model.add_transition( i1, m2, 0.15 )\n",
    "\n",
    "model.add_transition( i2, i2, 0.70 )\n",
    "model.add_transition( i2, d3, 0.15 )\n",
    "model.add_transition( i2, m3, 0.15 )\n",
    "\n",
    "model.add_transition( i3, i3, 0.85 )\n",
    "model.add_transition( i3, model.end, 0.15 )\n",
    "\n",
    "# Create transitions from delete states\n",
    "model.add_transition( d1, d2, 0.15 )\n",
    "model.add_transition( d1, i1, 0.15 )\n",
    "model.add_transition( d1, m2, 0.70 ) \n",
    "\n",
    "model.add_transition( d2, d3, 0.15 )\n",
    "model.add_transition( d2, i2, 0.15 )\n",
    "model.add_transition( d2, m3, 0.70 )\n",
    "\n",
    "model.add_transition( d3, i3, 0.30 )\n",
    "model.add_transition( d3, model.end, 0.70 )\n",
    "\n",
    "# Call bake to finalize the structure of the model.\n",
    "model.bake()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now lets try to align some sequences to it and see what happens!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence: 'ACT'  -- Log Probability: -0.513244900357 -- Path: M1 M2 M3\n",
      "Sequence: 'GGC'  -- Log Probability: -11.0481012413 -- Path: I0 I0 D1 M2 D3\n",
      "Sequence: 'GAT'  -- Log Probability: -9.12551967402 -- Path: I0 M1 D2 M3\n",
      "Sequence: 'ACC'  -- Log Probability: -5.08795587886 -- Path: M1 M2 M3\n"
     ]
    }
   ],
   "source": [
    "for sequence in map( list, ('ACT', 'GGC', 'GAT', 'ACC') ):\n",
    "    logp, path = model.viterbi( sequence )\n",
    "    print \"Sequence: '{}'  -- Log Probability: {} -- Path: {}\".format(\n",
    "        ''.join( sequence ), logp, \" \".join( state.name for idx, state in path[1:-1] ) )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first and last sequence are entirely matches, meaning that it thinks the most likely alignment between the profile ACT and ACT is A-A, C-C, and T-T, which makes sense, and the most likely alignment between ACT and ACC is A-A, C-C, and T-C, which includes a mismatch. Essentially, it's more likely that there's a T-C mismatch at the end then that there was a deletion of a T at the end of the sequence, and a separate insertion of a C.\n",
    "\n",
    "The two middle sequences don't match very well, as expected! G's are not very likely in this profile at all. It predicts that the two G's are inserts, and that the C matches the C in the profile, before hitting the delete state because it can't emit a T. The third sequence thinks that the G is an insert, as expected, and then aligns the A and T in the sequence to the A and T in the master sequence, missing the middle C in the profile.\n",
    "\n",
    "By using deletes, we can handle other sequences which are shorter than three characters. Lets look at some more sequences of different lengths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence: 'A'  -- Log Probability: -5.40618101242 -- Path: M1 D2 D3\n",
      "Sequence: 'GA'  -- Log Probability: -10.8868199358 -- Path: I0 M1 D2 D3\n",
      "Sequence: 'AC'  -- Log Probability: -3.62447187905 -- Path: M1 M2 D3\n",
      "Sequence: 'AT'  -- Log Probability: -3.64488075068 -- Path: M1 D2 M3\n",
      "Sequence: 'ATCC'  -- Log Probability: -10.6743329646 -- Path: M1 D2 M3 I3 I3\n",
      "Sequence: 'ACGTG'  -- Log Probability: -10.3938248352 -- Path: M1 M2 I2 I2 I2 D3\n",
      "Sequence: 'ATTT'  -- Log Probability: -8.67126440175 -- Path: M1 I1 I1 D2 M3\n",
      "Sequence: 'TACCCTC'  -- Log Probability: -16.9034517961 -- Path: I0 I0 I0 I0 D1 M2 M3 I3\n",
      "Sequence: 'TGTCAACACT'  -- Log Probability: -16.4516996541 -- Path: I0 I0 I0 I0 I0 I0 I0 M1 M2 M3\n"
     ]
    }
   ],
   "source": [
    "for sequence in map( list, ('A', 'GA', 'AC', 'AT', 'ATCC', 'ACGTG', 'ATTT', 'TACCCTC', 'TGTCAACACT') ):\n",
    "    logp, path = model.viterbi( sequence )\n",
    "    print \"Sequence: '{}'  -- Log Probability: {} -- Path: {}\".format(\n",
    "        ''.join( sequence ), logp, \" \".join( state.name for idx, state in path[1:-1] ) )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Again, more of the same expected. You'll notice most of the use of insertion states are at I0, because most of the insertions are at the beginning of the sequence. It's more probable to simply stay in I0 at the beginning instead of go from I0 to D1 to I1, or going to another insert state along there. You'll see other insert states used when insertions occur in other places in the sequence, like 'ATTT' and 'ACGTG'.\n",
    "Now that we have the path, we need to convert it into an alignment, which is significantly more informative to look at."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence: A, Log Probability: -5.40618101242\n",
      "ACT\n",
      "A--\n",
      "\n",
      "Sequence: GA, Log Probability: -10.8868199358\n",
      "-ACT\n",
      "GA--\n",
      "\n",
      "Sequence: AC, Log Probability: -3.62447187905\n",
      "ACT\n",
      "AC-\n",
      "\n",
      "Sequence: AT, Log Probability: -3.64488075068\n",
      "ACT\n",
      "A-T\n",
      "\n",
      "Sequence: ATCC, Log Probability: -10.6743329646\n",
      "ACT--\n",
      "A-TCC\n",
      "\n",
      "Sequence: ACGTG, Log Probability: -10.3938248352\n",
      "AC---T\n",
      "ACGTG-\n",
      "\n",
      "Sequence: ATTT, Log Probability: -8.67126440175\n",
      "A--CT\n",
      "ATT-T\n",
      "\n",
      "Sequence: TACCCTC, Log Probability: -16.9034517961\n",
      "----ACT-\n",
      "TACC-CTC\n",
      "\n",
      "Sequence: TGTCAACACT, Log Probability: -16.4516996541\n",
      "-------ACT\n",
      "TGTCAACACT\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def path_to_alignment( x, y, path ):\n",
    "    \"\"\"\n",
    "    This function will take in two sequences, and the ML path which is their alignment,\n",
    "    and insert dashes appropriately to make them appear aligned. This consists only of\n",
    "    adding a dash to the model sequence for every insert in the path appropriately, and\n",
    "    a dash in the observed sequence for every delete in the path appropriately.\n",
    "    \"\"\"\n",
    "    \n",
    "    for i, (index, state) in enumerate( path[1:-1] ):\n",
    "        name = state.name\n",
    "        \n",
    "        if name.startswith( 'D' ):\n",
    "            y = y[:i] + '-' + y[i:]\n",
    "        elif name.startswith( 'I' ):\n",
    "            x = x[:i] + '-' + x[i:]\n",
    "\n",
    "    return x, y\n",
    "\n",
    "for sequence in map( list, ('A', 'GA', 'AC', 'AT', 'ATCC', 'ACGTG', 'ATTT', 'TACCCTC', 'TGTCAACACT') ):\n",
    "    logp, path = model.viterbi( sequence )\n",
    "    x, y = path_to_alignment( 'ACT', ''.join(sequence), path )\n",
    "    \n",
    "    print \"Sequence: {}, Log Probability: {}\".format( ''.join(sequence), logp )\n",
    "    print \"{}\\n{}\".format( x, y )\n",
    "    print"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In addition to getting this alignment, we can do some interesting things with this model! Lets score every sequence of length 5 of less and see what the distribution looks like."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAmgAAAFKCAYAAACpY6PpAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3Xl4VPW9P/D3mTUzk0lmJpmZLIQQArIEVFxSEVvUUhWX\ntrcVjVusWntvvVr6QK+l0ovaarGLl4ty/bXPba22V5tWlCq2SBUVpQEBF8CwJSH7Nksm2ySZ9fz+\nmEwEgTBAzpwzk/fr0YdMzpwzn3CY5J3vKoiiKIKIiIiIFEMldwFEREREdCwGNCIiIiKFYUAjIiIi\nUhgGNCIiIiKFYUAjIiIiUhgGNCIiIiKF0Uj9AqtXr8aePXsgCAIeeughzJ07d/TYjh07sGbNGqjV\napSUlODxxx/Hzp07sXTpUkyfPh2iKGLGjBn48Y9/LHWZRERERIohaUDbtWsXmpqaUFVVhfr6eqxc\nuRJVVVWjxx9++GH88Y9/hMPhwNKlS/Hee+8hIyMD5eXlWLt2rZSlERERESmWpF2c27dvx6JFiwAA\npaWl6Ovrg9/vHz3+yiuvwOFwAABsNht6enoAAFw7l4iIiCYySQOax+OBzWYbfWy1WuHxeEYfm0wm\nAIDL5UJ1dTUWLlwIAKivr8d9992H2267DdXV1VKWSERERKQ4ko9BO9qJWsa8Xi+++93v4pFHHkF2\ndjaKi4tx//33Y/HixWhpaUFlZSXefPNNaDRJLZWIiIhINpK2oDkcjmNazFwuF+x2++jjgYEB3Hvv\nvVi2bBnmz58PAHA6nVi8eDEAoKioCLm5uejq6hrzddglSkREROlE0mapBQsWYN26dbjppptQU1MD\np9MJo9E4evyJJ57AXXfdhQULFox+buPGjXC73bj77rvhdrvh9XrhdDrHfB1BEOB290v2dZB07HYz\n710K4/1Lbbx/qYv3LrXZ7eZTPkfSgDZv3jyUlZWhoqICarUaq1atwoYNG2A2m3HZZZfhtddeQ3Nz\nM/7yl79AEATccMMNuO6667Bs2TJs2bIF4XAYjz76KLs3iYiIaEIRxDTpH+RvEqmJvwWmNt6/1Mb7\nl7p471JbIi1o3EmAiIiISGEY0IiIiIgUhgGNiIiISGEY0IiIiIgUhgGNiIiISGEY0IiIiIgUhgGN\niIiISGEY0IiIiIgUhgGNiIiISGEY0IiIiIgUhgGNiIiISGEY0IiIiIgUhgGNiIiISGEY0IiIiIgU\nhgGNiIiISGEY0IiIiIgUhgGNSKGioih3CUREJBON3AUQ0fEGh0P46R8+hE6jwp3XzMTUgiy5SyIi\noiRiCxqRAr34Vi26ugfR4hrA43/cjaottQiEInKXRUREScKARqQwH9e6Uf1pJ/Kseiy5vBRZRi3+\nsasF617eK3dpRESUJAxoRAoyMBTC828cglol4CsXFaIkPwt3XzsbBTlG1DT64O4ZkrtEIiJKAgY0\nIgV54c3D6PMHMX+WFfm52QAArUaFc0tzAQAf1HTKWR4RESUJAxqRQrh6hvDB/i7kWfX4QlnRMcem\nT8qGSgC213TIVB0RESUTAxqRQnx0yA0AmF2cBbX62LemQa/BlPwsdHQPo8Prl6M8IiJKIgY0IoX4\n8LALggDMLHac8PisyVYAwA62ohERpT0GNCIF8PUHUN/Wh6JcAzKNuhM+Z9qkbKhVAnbUdEHkIrZE\nRGmNAY1IAT6ujXVvTskznfQ5eq0apQVZcPcG0OZmNycRUTpjQCNSgA9Hxp/NLM4d83kzi2PdnNWf\ntkteExERyYcBjUhm/YNBHGruQYEtAxazYcznlhZkQ6tRYddBV5KqIyIiOTCgEcnskzoPoqKIknzj\nKZ+r1ahQmGuCty8I/3AoCdUREZEcGNCIZBZfXuOcyWN3b8Y5rbEg19TZL1lNREQkLwY0IhkNB8Oo\naeyGw6KH3XLyCQJHy8uJBbTalm4pSyMiIhkxoBHJqK61F+GIiMn2sceeHS3PFgto9W29UpVFREQy\nY0AjktHB5h4AwCRHZsLnZBm1MOjUaHUPSlUWERHJjAGNSEaHWnxQCUBxvi3hcwRBQJ7NiB5/CAND\nnChARJSOGNCIZBIIRtDY0Y88awb0WvVpneu0caIAEVE6Y0AjkkldWy8iURH5ORmnfW58HNrhZu94\nl0VERArAgEYkk0MtPgBAoT3x8Wdx8YB2pJ0TBYiI0hEDGpFMDjX3QBCAkoLEx5/FmY1aGPUatHo4\nUYCIKB0xoBHJIBCK4Eh7H5wW/WmPPwM+myjQ6w+jfzAoQYVERCQnBjQiGRwZGX9WeAbjz+I4UYCI\nKH0xoBHJIL7+WYHdfMbXyLPFFrc9zB0FiIjSDgMakQwOtfRAAFBSYD3ja4xOFOCOAkREaYcBjSjJ\nQuGR8WdWPTJ0mjO+TqZBC1MGJwoQEaUjBjSiJKtv60M4EkW+7czHnwGxiQJ2iwF9g2EMBcLjVB0R\nESkBAxpRkh1qGRl/lms662vZsmIhr7ObrWhEROmEAY0oyQ41+yAAmFqYc9bXysnSAwDaPQNnfS0i\nIlKOMx8Ak6DVq1djz549EAQBDz30EObOnTt6bMeOHVizZg3UajVKSkrw+OOPn/IcolQWCkdR394H\nu0UPg/7s3342c6wFraWzF5hbcNbXIyIiZZA0oO3atQtNTU2oqqpCfX09Vq5ciaqqqtHjDz/8MP74\nxz/C4XBg6dKleO+992AwGMY8hyiVNXT0IRSOouAsx5/F2eItaF7/uFyPiIiUQdIuzu3bt2PRokUA\ngNLSUvT19cHv/+wHySuvvAKHwwEAsNls6OnpOeU5RKnsUHN8/82zH38GxGZyajUquHoC43I9IiJS\nBkkDmsfjgc322T6DVqsVHo9n9LHJFPsh5XK5UF1djYULF57yHKJUFl+gdsoZ7L95IoIgwJalR3d/\nENGoOC7XJCIi+SV1koAoHv8DxOv14rvf/S4eeeQRZGdnJ3QOUSoKR6Kob+uFPVsHU4Z23K5rM2cg\nHBHh7Rset2sSEZG8JB2D5nA4jmn9crlcsNvto48HBgZw7733Yvny5Zg/f35C55yM/Sy2zCF5TZR7\nt7/Bi2A4ipICMywW47hdt9BhxoEmHwZDUVn+LifK/UtXvH+pi/cuvUka0BYsWIB169bhpptuQk1N\nDZxOJ4zGz34wPfHEE7jrrruwYMGChM85GbebG0anIrvdPGHu3Qd72wEAOZk69PSM37plRl2sIfzT\n2i4Uj9PYtkRNpPuXjnj/UhfvXWpLJFxLGtDmzZuHsrIyVFRUQK1WY9WqVdiwYQPMZjMuu+wyvPba\na2hubsZf/vIXCIKAG264AUuWLMHs2bOPOYcoHcQXqC0pHJ/xZ3HxpTZaXX3jel0iIpKP5OugLVu2\n7JjHM2bMGP147969Jzxn+fLlktZElGzhSBR1rb3IzdIh06Ab12tbzbGlNrp8HINGRJQuuJMAURI0\ndvQjEIqgIGd81j87mlajQrZJB3cvl9ogIkoXDGhESbCnPjbxZbIjU5Lr27L08A9HMDjMTdOJiNIB\nAxpREuyr90KtElBadPb7b54IN00nIkovDGhEEvP1B9DsGkCR3QC9Vpphn/GJAm2c1UVElBYY0Igk\ntu+IFwAw2WGQ7DVyRvbkbOnqkew1iIgoeRjQiCS2py42/qx0kjTdm8BnXZztHnZxEhGlAwY0IgmF\nwlHsb/TBZtbBbpFuEVlThgY6rYozOYmI0gQDGpGEDrf2IBCKoFjC7k1gZNN0cwa6+0PcNJ2IKA0w\noBFJaG9dbPzZlPwsyV/LatYjEhXR3c8Fa4mIUh0DGpGE9tZ7oNUIKCkY3+2dTsSSGZso4PINSf5a\nREQkLQY0Iol0dQ+iyzeEYocRGrX0bzVrZmwLqXYutUFElPIY0Igk8v7eDgBASZ50kwOOZhnZk7Pd\nw4BGRJTqGNCIJBAKR/DennYY9GrMKc1Lymuyi5OIKH0woBFJYOcBFwaGQigrNkOrUSflNU0ZGmg1\nKnj7g0l5PSIikg4DGtE4E0URb33YCkEAzp/mSNrrCoIAS6YOvv4gRJFLbRARpTIGNKJxdqSjD02d\n/SjNN8GWnZzxZ3GWTD2CYRF9g6Gkvi4REY0vBjSicfb2h20AgDkl1qS/dnwcmpvj0IiIUhoDGtE4\n6vUHsetgF3LMOkwvyk3661tHZnJ2eAeS/tpERDR+GNCIxkk0KuLZvx1AOCJibkkWBEFIeg2W0bXQ\n+pL+2kRENH4Y0IjGyV+3HcG+I16UOA24cGahLDXEuzg7uwdleX0iIhofDGhE4+DDQ268Xt0Ea6YW\niy8pgToJOwecSJZRB5VKgKePS20QEaUyjdwFEKWyYCiCHfu78KcttdBqVLj2C5OQadTLVo9KJSDb\npEM310IjIkppDGhECerzB9Hh9WNgKITB4TDaPH78c18H/MNhqATgmovzUOiwyF0mLJk6NHQEMBQI\nw6DnW5yIKBXxuzfRGGoau7Fldyuauvrh6w8cd9yoV6N8hgXnTXPCmmWUocLjWTP1aEA/XL4hFOeZ\n5S6HiIjOAAMa0QkMDofxl3dq8d6e2IbnmQYNSvNNsGZqYdBroNeqYTLoMLXQmrStnBIVnyjQ1e1n\nQCMiSlEMaESfU9fai//36qfw9QfgyNbhyxcUoCgv+YvOninLyFpore4+lCM5G7UTEdH4YkAjOoqr\nZwhr1+/BUDCCS2ZacencImgU1kJ2Ktb4Uhtev8yVEBHRmWJAIxoxFAjj6fV74R8O48vz7Lhw5iS5\nSzoj2SOL1Xp6jx8zR0REqYHroBEBiIoi/nfjfrR5/Dh/albKhjMA0KhVMBu1XGqDiCiFMaARAXi9\nuhGf1HlQ7DDgyxeVyF3OWbNk6tE3GEYoHJG7FCIiOgMMaDThdXUP4vXqRpiNGlw7f4psuwCMp/hM\nTnfPsMyVEBHRmUj9n0REZ0EURfzfm4cRjoj44pxcmI0Zcpc0Lqwj49BcPu7JSUSUihjQaEL78JAb\nNQ3dmOI0oGxq+ixJEV9qo83dJ3MlRER0JhjQaMIaDobxpy21UKsELDyvAIIgyF3SuIl3cXZ4BmSu\nhIiIzgQDGk1Yr1c3wdcfwIXTsuHMyZK7nHEVXwvNzaU2iIhSEgMaTUjdfcN4c3cLsowazJ9bJHc5\n406vU8Og18Dbx6U2iIhSEQMaTUiv/bMRoXAU5TNs0OvSc71mS6YOPf4golFR7lKIiOg0MaDRhNPZ\nPYhtezuQk6XD+ecUyF2OZCyZekSjsdZCIiJKLQxoNOFseO8IoqKIS2bmQKVKn4kBn2cdmcnp6hmS\nuRIiIjpdDGg0oTR19mPXQRfybXrMnuqUuxxJWUbWQmt398tcCRERnS4GNJpQXn6vHgBwyWx7Wi2r\ncSLxpTbaPQxoRESphgGNJoy61l58eqQbkx0GTC+yy12O5OJLbbh87OIkIko1DGg0Yfx12xEAQPnM\n9A9nAGDM0ECrUcHDpTaIiFIOAxpNCIdberC/0YdihwFTC3PkLicpBEGAJVMH30AQosilNoiIUgkD\nGk0If31/pPVs1sRoPYuzZuoRCovoGwzJXQoREZ0GBjRKe4eafTjY3IMSpxElBROj9SzOMjoObVDm\nSoiI6HQwoFHae3VbAwCgfJZD5kqSzzKyFlqn1y9zJUREdDok3+Nm9erV2LNnDwRBwEMPPYS5c+eO\nHgsGg1i1ahVqa2vx8ssvAwB27tyJpUuXYvr06RBFETNmzMCPf/xjqcukNFXX1htrPcszojjfKnc5\nSRefydnu7gNQKG8xRESUMEkD2q5du9DU1ISqqirU19dj5cqVqKqqGj3+i1/8ArNmzUJdXd0x55WX\nl2Pt2rVSlkYTxKYdTQCAC6bnylyJPOKL1XZ2s4uTiCiVSNrFuX37dixatAgAUFpair6+Pvj9n3W1\nLFu2bPT40TjjjMZDm3sAH9d6UJiTgamFNrnLkYXZqINKJcDTF5C7FCIiOg2SBjSPxwOb7bMfjFar\nFR6PZ/Sx0Wg84Xn19fW47777cNttt6G6ulrKEimNbfqgGQAwb5ot7XcNOBmVSkC2SYfufs7iJCJK\nJZKPQTtaIi1jxcXFuP/++7F48WK0tLSgsrISb775JjSasUu1283jVSYlmRT3ztU9iA/2d8FpzcAl\n86ZANUEDGgDYrUYcbvbBmJkBk0E7/tfney+l8f6lLt679CZpQHM4HMe0mLlcLtjtY69D5XQ6sXjx\nYgBAUVERcnNz0dXVhcLCsQc4u7khdEqy282S3LsX3zyMSFTEuSVZ6Oud2FsdmTNib/OaWhdK8rPG\n9dpS3T9KDt6/1MV7l9oSCdeSdnEuWLAAmzdvBgDU1NTA6XQe160piuIxLWsbN27Es88+CwBwu93w\ner1wOp1SlklpZigQxra9HcgyajB3Wr7c5cjOlhWfyTkgcyVERJQoSVvQ5s2bh7KyMlRUVECtVmPV\nqlXYsGEDzGYzFi1ahKVLl6KzsxONjY2orKzEzTffjCuvvBLLly/Hli1bEA6H8eijj56ye5PoaB8c\n6EIgFMEFpVZo1Fzqz2bOAAC0dPUCKJC3GCIiSojkyWfZsmXHPJ4xY8boxydbSuPXv/61pDVRenvv\nk3YIAjCnlC2vwFEtaFyslogoZbB5gdJKc1c/Gjv7MTXPCGuWQe5yFCHToIVWo4Krh0ttEBGlCgY0\nSitb97QDAGZNzpa5EuUQBAFWsx7d/UFEucYgEVFKYECjtBEIRbCjpgtmgwYzp7B782g2sx7hiIju\nvmG5SyEiogQwoFHa2H3QhaFAGLOKMqFSTdx1z07ElhWbKNDh5ZZPRESpgAGN0sZ7I92bc6c5ZK5E\neWzm2ESBVlefzJUQEVEiGNAoLfj6A6ht7cVkuwE52Sa5y1GceAtaq4sLWxIRpQIGNEoLe+pjO1ZM\nyTvx/q4TXbwFrcs3sXdVICJKFQkFtF/96ldobGyUuBSiM/dJbSygTS/KlbkSZdJp1cg0aOHu5VIb\nRESpIKGAlp2djeXLl+OOO+7AX//6VwQC/CZPyhEIRrC/0Qd7tg452WxBOxlblh79g2EEghG5SyEi\nolNIKKDde++9ePnll/Gzn/0MLpcLd955Jx555BHU19dLXR/RKdU0diMciWKKk+FsLPEtnzq7OZOT\niEjpTmsMWmdnJ5qamuD3+2EymbBixQq8+OKLUtVGlJB492ZpoUXmSpRtdMsnDzdNJyJSuoT24ly3\nbh1ee+01TJkyBTfffDN+8pOfQK1WIxgM4sYbb8Stt94qdZ1EJxSNithT74EpQ40iJwPaWHKyPts0\nff6cfJmrISKisSQU0DweD37/+9+jsLBw9HMtLS0oKirCD37wA8mKIzqVIx196B8MYc6ULAgCF6cd\nS3wmJzdNJyJSvlN2cUajUdTX16OgoADRaBTRaBTBYBD33XcfAOBLX/qS5EUSnUy8e3OKM1PmSpQv\ny6SDRi3AzU3TiYgUb8wWtNdffx1PP/00mpqaMGvWrNHPq1QqXHbZZZIXR3Qqn9R5oFULmD7ZLncp\niicIAqyZenj7AhBFkS2OREQKNmZAu/7663H99dfj6aefxgMPPJCsmogS0jMQQLvHj6l5Rmg1XHM5\nEbasDLh7h+HrD4zuLkBERMozZkDbunUrFi5ciLy8PKxfv/644zfeeKNkhRGdSl1rLwDAaWXQSFRO\ndgbQArS6BxjQiIgUbMyAdujQISxcuBAfffTRCY8zoJGcakcC2iSHWeZKUofTagAAHGnz4dxS7rpA\nRKRUYwa073znOwCA1atXj45ZCQaD8Hq9yM/nNH2SV11bD1QqYJKDy2skymGJBbSGjj6ZKyEiorEk\ntMzGb37zGxiNRixZsgTf+MY3YDKZcNlll2Hp0qVS10d0QoFgBM1dA8izZnD82WnIMumg16rR7uWm\n6URESpbQT7Z33nkHt99+OzZt2oQrrrgCL730Ej788EOpayM6qSMdfYhEReRZ9XKXklIEQYDdYoC3\nL8g9OYmIFCyhgKbRaCAIAt577z0sWrQIQGx9NCK51LX2AADyc0wyV5J64uPQWt3c8omISKkSCmhm\nsxnf+c53UF9fj3nz5uGdd97hGkokq9q22ASB4nyrzJWkHoc1Pg6tV+ZKiIjoZBIag/bkk0+iuroa\nF1xwAQBAp9Ph5z//uaSFEZ1MNCqivq0XNrMWmQad3OWknPhEgSOtPuCiyTJXQ0REJ5JQQFOr1QBi\nY9FEUQQAdHR0cJkNkkWbx4+hQARTnUa5S0lJOdkZUAlAq2dQ7lKIiOgkEgpo99xzD1Qq1TGbpQNc\nB43kER9/5rRxodUzoVGrkJOdgS7fMKJRESoVhysQESlNQgEtHA6jqqpK6lqIEhIffzY5j+PPzpTD\nYoS7ZxhdvkFOtCAiUqCEJglMmzYNPp9P6lqIElLb0guDXg27hV2cZyo+UaCps1/mSoiI6EQSakHr\n7OzEVVddhdLS0tHxaADwwgsvSFYY0Yn0+oPw9g1jar6RM4nPQjyg1bd245KyPJmrISKiz0sooMW3\nfCKSW4sr1uKTa+bszbMRn8nZ1MW10IiIlCihLs7y8nIMDg7i8OHDKC8vR15eHi6++GKpayM6TstI\noMhl9+ZZMeg1yDJq0dHNLZ+IiJQooYD2y1/+EuvXr8crr7wCANi4cSMee+wxSQsjOpEWVyygFdiz\nZa4k9TmsRviHI+gdCMhdChERfU5CAW3Xrl1Yt24dTKbYbK9///d/R01NjaSFEZ1Is2sAOo0KVjP3\n4Dxb8S2f6tq4owARkdIkFND0+mN/GEYiEUQi3GiZkisYiqDTOwi7RccJAuNgstMMANhb55K5EiIi\n+ryEJglccMEFWLFiBdxuN37/+99j8+bNKC8vl7o2omO0efyIiiJyOEFgXBTkGKFVq3CwmS1oRERK\nk1BA++pXv4pDhw5h3759+Oijj3DPPffgK1/5itS1ER0jPv4sJ4vdm+NBrVZhkt2Ehs5+9AwEYMnk\n3ysRkVKMGdCGh4exfPlyHDx4EHPmzIHT6cTu3buh1+uxcOFC6HRsyaDkic/gzMvNkrmS9DE5z4yG\nzn7sb+jGpXPz5S6HiIhGjDkG7ZlnnkF+fj42b96MtWvX4tlnn8Xbb7+NjIwMrFmzJlk1EgEAml39\nEAQgz5Ypdylpo3hkHNqeui6ZKyEioqONGdB2796NBx98EBrNZw1tBoMBDz/8MLZt2yZ5cURxUVFE\ni2sAtkwdtJqE5rZQApxWAzJ0ahxu6YMoinKXQ0REI8b8SadWq0/YjanVapGVxW4mSh5P7zCGgxHk\nZrNbfTwJgoDJTjN6B8Nw9XDRWiIipRgzoI21lMHRe3ISSS0+/owzOMdfvJuz5ohX5kqIiChuzEkC\nH3/8MS6//PLjPi+KInw+n1Q1ER0nvgenw2qSuZL0U5wXG9O3p86FKy8skrkaIiICThHQ3njjjWTV\nQTSm+BIb+XZ2rY83a6YeZqMWde0DiIoiVFwEmIhIdmMGtMLCwmTVQTSm5q4BmDLUMBvZxTneBEFA\nsdOMTxu60dzVjyl5DMFERHLjdDhSvMHhELx9w7BncyFVqUyfZAEAvLmrWeZKiIgIYECjFNDhHQQA\nWDO1MleSvkoLsmDJ1GPnATd6/cEzukYwFMEntR78+e1a7KnzjHOFREQTS0JbPRHJqd3rBwBYMtm9\nKRWVSsDFM+14c3crtuxuwTcWliZ8bu9AAC+8VYu9dR4Ew1EAwOadLbikzIlbF52DTAODNRHR6ZK8\nBW316tWoqKjALbfcgn379h1zLBgMYsWKFfjmN7+Z8Dk08XSOtKDZrdxBQEplJTZk6NR4+6NWBEOR\nhM5p6ujDY3/Yjd0HXTAbNLhwugU3zJ8EpzUDO2q68OP/3YFDzZzxTUR0uiQNaLt27UJTUxOqqqrw\n2GOP4fHHHz/m+C9+8QvMmjXrmPXWTnUOTTzxLk4usSEtnUaN86flYjAQQfWnnad8/r4jXvzH0+/D\n2xfApbNtuPu62fjyRSWYNcWOO66aiS+dl4+BoRB+81oNQuHEAh8REcVIGtC2b9+ORYsWAQBKS0vR\n19cHv98/enzZsmWjxxM9hyaeju5BGHQqGPTskZfavOl2qAQBf9/RiOgYWz+9/VEr/vulPQiFI1h8\ncR4uO6/4mF+0VCoBl8zOw0UzHegZCOKt3a3JKJ+IKG1IGtA8Hg9sNtvoY6vVCo/ns8HDRqPxtM+h\niSUcicLtG4LNrBtzZwsaH2ajFrOnWOHpDeDld+sRjR4b0qJREX96qxb/94/DMOrVqLxmOuZOyz/p\n9S6Z7USGTo2N1Q0YGApJXT4RUdpIapPEmWzGnOg5drv5tK9NyjDWvWvu7ENUFOGwGWCxHB/oafwt\nXlCCdo8fmz5ohrt3GMtvvwgGnRqHm3vwly2HsftAF5zWDFR8ZRqcOdmnvN4VFxZh0/ZGbPm4Dd/+\n2lzpvwA6Lfzembp479KbpAHN4XAc0/rlcrlgt9vH/RwAcLv7z7xQko3dbh7z3u2vcwMAjDoVenoG\nk1XWhCYAuP2qc/DqtgbsPujCd594C4FQBEOB2DiyYocB1186FXp1bHbmqe7LrKJs/HOPFq9va8CC\n2U7kWgxSfwmUoFO9/0i5eO9SWyLhWtIuzgULFmDz5s0AgJqaGjidzuO6NUVRPKaVLJFzaOKITxDI\nyea/gWTK0Gmw5PJpKJ/pQK8/iAytCueWZOG68gLcePk5MBkSX/JEo1bhi+cWIBIV8df36yWsmogo\nfUjagjZv3jyUlZWhoqICarUaq1atwoYNG2A2m7Fo0SIsXboUnZ2daGxsRGVlJW6++WZcd911mD17\n9jHn0MTVMbIGmtPGpvxkU6kEXD6vEF88rwBq1dmN/5s9xYpt+zqw+5AblddEoNOqx6lKIqL0JPkY\ntGXLlh3zeMaMGaMfr1279oTnLF++XNKaKHV0eAehVgnINnGRWrmcbTgDYvt9zpxswQcHXNh3xIsL\nZzjGoTIiovTFrZ5IsURRREf3IGxmLVTjEBJIXjMnWwEA2/ZyyQ0iolNhQCPF6hkIIhCMwMI9ONOC\nw2qA1axHTWMvhoNhucshIlI0BjRSrNE9ONm9mRbi3ZzhiIhParm2IRHRWBjQSLE6R2dwclmGdDGr\nONbNWb3W3TibAAAfvUlEQVSvTeZKiIiUjQGNFCs+g9PBTdLTRm62AbnZGTjQ3IehALs5iYhOhgGN\nFCu+BpotO0PmSmg8zZxsRSQq4qPDbrlLISJSLAY0UqwOrx9ZRg10Gq6ZlU5mTrYAYDcnEdFYGNBI\nkYYCYfQMBGE1c4JAurFlZSAnKwN1bf0IhaNyl0NEpEgMaKRInd2x7k2LiUtspKMpeWaEIiLq23rl\nLoWISJEY0EiR2j2xCQJWs17mSkgKxXmxrbs+qe2SuRIiImViQCNFiregcQZneipyZEIlADUN3XKX\nQkSkSAxopEjxGZy5FqPMlZAU9Fo1CnJNaPMOwz8ckrscIiLFYUAjRerw+qHXqmDK0MhdCkkk3s25\nv5GtaEREn8eARooTjkTh8g3BZtZBELhJerqakpcFAPjkMMehERF9HgMaKY67ZwiRqAgrN0lPa/k2\nI/RaNQ42cyYnEdHnMaCR4sT34LQwoKU1lUrAZGcmfAMhuHqG5C6HiEhRGNBIcTpGZnDasjhBIN1N\nGRmH9mm9R+ZKiIiUhQGNFKdjZA00p41LbKS7YmcsoO2pc8lcCRGRsjCgkeJ0dA9CpQIsZm6Snu6s\nZj2yjFrUtfUjKopyl0NEpBgMaKQooiiiwzsIa6YOahVncKY7QRBQ5DBjKBhFu9svdzlERIrBgEaK\n0ucPYigQ5gzOCWSyM9aVvb/RK3MlRETKwYBGitLOGZwTTpEjFtD21btlroSISDkY0EhROr0jm6Rn\ncvzZRJFt0iHLqMWRDj/HoRERjWBAI0WJ78Hp4AzOCeOzcWgRjkMjIhrBgEaKEl8DjZukTyzxbk6O\nQyMiimFAI0Xp8PphNmig16rlLoWSKD5R4NMjXLCWiAhgQCMFGQ6G0d0XgM3MCQITTbZJB7NRiyMd\nAxyHRkQEBjRSkK7u2H6MnME58QiCgMmOTAwGImj3cBwaEREDGilGx8gMzmwTA9pEVOSIbfu0v4Hj\n0IiIGNBIMeJroOVaTDJXQnLgODQios8woJFixNdAc1i5xMZEFB+HVs9xaEREDGikHB3dg9BpVDAb\ndXKXQjIQBAHFTjOGAhG0ugbkLoeISFYMaKQI0aiIru5B2MxaCAI3SZ+oivNi49D21bObk4gmNgY0\nUgR37xDCERFWM1vPJrJiZyyg7eW+nEQ0wTGgkSLEt3iycAbnhJZp0CI3OwMNnX6EwlG5yyEikg0D\nGilC50hAs5q5SfpEV5xnRjgior6tV+5SiIhkw4BGihBfA81hM8tcCcltykg35546l8yVEBHJhwGN\nFKHDOwhBAHKy2II20U1yZEIlADUN3XKXQkQkGwY0kp0oiujw+mHN1EKt5j/JiU6vVaMg14Q27xAG\nh0Nyl0NEJAv+NCTZ9Q+F4B8Ow5rJGZwUU+w0QxSBA00+uUshIpIFAxrJrpMzOOlz4uuh7antkrkS\nIiJ5MKCR7NpHJghwDTSKy88xQatR4UATZ3IS0cTEgEayi7eg5XIPThqhVgmYkmeGtz84OsOXiGgi\nYUAj2cUXqXVYTTJXQkpyziQLAGDn/k6ZKyEiSj4GNJJdh9cPU4YaGTqN3KWQgkwrzIZKJWDnAY5D\nI6KJhwGNZBUIReDtHYaN48/oc/Q6NabkmdHRPQyXb1DucoiIkkrygLZ69WpUVFTglltuwb59+445\nVl1djSVLlqCiogLPPPMMAGDnzp2YP38+Kisrcccdd+Cxxx6TukSSUbt7ACIASyZncNLxRrs52YpG\nRBOMpH1Ku3btQlNTE6qqqlBfX4+VK1eiqqpq9Pjjjz+OZ599Fg6HA7fffjuuvvpqAEB5eTnWrl0r\nZWmkEE2d/QAAK5fYoBOYPikbm3fFxqFdf2mJ3OUQESWNpC1o27dvx6JFiwAApaWl6Ovrg98fm5HV\n0tICi8UCp9MJQRCwcOFC7NixA0BsZXmaGBrbY8so2G2cwUnHM+g1mOwwo9UzBG/vsNzlEBEljaQB\nzePxwGazjT62Wq3weDwnPGaz2eByxTZHrq+vx3333YfbbrsN1dXVUpZIMmvs6AMAFORmyVwJKdWM\nybFuzt0H2c1JRBNHUicJjNUyFj82ZcoU3H///XjmmWfwxBNPYOXKlQiHw8kqkZKsqaMPZoMGBj1n\ncNKJTS/MBgDs2N8hcyVERMkj6U9Fh8Mx2mIGAC6XC3a7ffSY2+0ePdbV1QWHwwGHw4HFixcDAIqK\nipCbm4uuri4UFhaO+Vp2u1mCr4Ck1D8YhKd3GOcUmWGxGOUuh86Q1PfOYgGmFmTjSHsvBiMiivPY\n2jqe+L0zdfHepTdJA9qCBQuwbt063HTTTaipqYHT6YTRGPtmXlhYCL/fj/b2djgcDrz77rt48skn\nsXHjRrjdbtx9991wu93wer1wOp2nfC23u1/KL4UkcKg5thG2Wa9GTw+XUUhFFosxKffu/Gk5ONLe\ni+df24fv/su5kr/eRGG3m/m9M0Xx3qW2RMK1pAFt3rx5KCsrQ0VFBdRqNVatWoUNGzbAbDZj0aJF\nePjhh7Fs2TIAwPXXX4/i4mLk5uZi+fLl2LJlC8LhMB599FFoNOz+SkctrgEAQE52hsyVkNKVFmTB\nnp2B3Yc9cPUMwWExyF0SEZGkBDFNpkzyN4nU89ymg3hvTzsqr5qOvBzO4kxFyWpBA4ADTT5srG7E\nZXOduPu6sjO6RlQUUdvSgwNNPhxs8qHF7ceUPDPOn5aL86bnTrjgx1aY1MV7l9pkb0EjGkurewBq\nlQA7x59RAmYUWfB+ph7ba1z4ly9Ng9WsP63zO7x+PL/pIA63xpZ2EQQg26jFgSYfDjT58KcttVj8\nhcn45uWlUAmCFF8CEVHCGNBIFlFRRKt7AA6rHmo1dxyjU1OpBFwy24k3djZj045G3PqVGQmdF45E\n8fftTXh9eyPCERHTCkyYOdmCkgIrDHot+geDqGvrxa6DXdj0QTN8AwHcfe0saPjvkohkxIBGsnD3\nDCEYisJu4fgzSlzZFCv+ua8D737Sji+U5aG0IHvM57t6hvCbVz9FQ0c/zAYNFp5rx+ypecc8x2zU\nYd50O2ZOtmL91jrsqOlCnz+If/+XuVz+hYhkw18RSRatrtiOErnZp9dNRRObWq3CoosmIRIR8V9/\n/gTNXScfg7PzQBce/f1ONHT0Y/bkTNx59YzjwtnRDHoNKq48B6UFWdjf6MOzf9vPXU2ISDYMaCSL\nVndsBmdeLicH0OmZPsmCay8pxlAggl/+6WO0efzHHD/S3oen1u/Fr1+tQSQi4qoLHbh+wXQYDbpT\nXlurUeFfvjgVhbkmfHjYg+pPO6X6MoiIxsT2e5JF68gSG0VOK8RwROZqKNWUldgQikTxj10t+Onz\nuzDJngmHxYBefxAHmmLr603KzcCXLyiEM+f0FrZVqQRcP78Yv990EP/3j0M4p8gC+wSb3UlE8mML\nGsmi1T0Ag16NLNOpWzWITuT8abm4urwIRr0ajZ192LG/CweafCh2GPDNLxbhlkUzTzucxWVn6rHo\nokkIhKL4zWufIhplVycRJRdb0CjpAsEIXL4hFNkNELicAZ2F80pzcV5pLqJREf1DIUSj4mkvv3Ey\nZVNsqG/rw6GWHmze2YzFlxSPy3WJiBLBFjRKujaPHyIAm5mtZzQ+VCoB2SbduIUzABAEAVddXIQM\nnRqv/bMBfYPBcbs2EdGpMKBR0jWNzLzLyeIMTlI2g16DBXPzEQhF8er7R+Quh4gmEAY0Srr6tthK\n7kVOi8yVEJ3a+aU5sJr12PpJOzq8/lOfQEQ0DhjQKOnq2/ug16pgt3KLJ1I+tVqFy88vQFQE/rzl\nsNzlENEEwYBGSTUwFEJX9yDybRmcIEApY1phNibZTdh7xIdDzT65yyGiCYABjZLqSHsfAMBh4fgz\nSh2CIOCKeYUAYq1o3GGAiKTGgEZJdaQ9Nv4sP8ckcyVEpyc/x4Tpk7LR2OXHpw3dcpdDRGmOAY2S\nqn6kBY0TBCgVLZiTDwBY/24tW9GISFIMaJQ0UVHEkfY+2MxaGDO0cpdDdNocVgNmFFnQ4hrEnnqv\n3OUQURpjQKOk6fAOYigQRp41Q+5SiM7Ygrl5AICX2YpGRBJiQKOkOTKy/hknCFAqy802YFaxFW2e\nIXxc65G7HCJKUwxolDTx8WeTHBx/Rqnt0jl5EAC8/G4domxFIyIJMKBR0hxp74VWLSAvJ1PuUojO\nSk5WBmZPsaKjewgfHXLLXQ4RpSEGNEqKoUAYbW4/8mwZUKm4QC2lvvlz8iAIwMtb6xCNshWNiMYX\nAxolRWNHH0QATo4/ozRhM2dgzhQbunzD2HXQJXc5RJRmGNAoKfY3xbbHyeMCtZRG5s/Jg0oAXmEr\nGhGNMwY0Soq99V6oVQKmFeXIXQrRuLFk6jFnag7cvQF8sL9L7nKIKI0woJHkfP0BtLgGUGQ3QKdR\ny10O0biaX5YHlSBg/bt1CIUjcpdDRGmCAY0kt7c+tlbUZLtB5kqIxl+2SYcLZ9jhGwhi884Wucsh\nojTBgEaS2zuyJU4puzcpTV1algeDXo3XqxvRMxCQuxwiSgMMaCSpUDiK/Y0+2Mw62C2cIEDpSa9T\n44vnFiAYjmL9O3Vyl0NEaYABjSR1uLUHgVAExQ52b1J6O3dqDuyWDFTXdKGps1/ucogoxTGgkaT2\n1sW6N4vzsmSuhEhaKpWAK+dNAgA8t2k/ItGozBURUSpjQCNJ7T3ihVYjYGqhTe5SiCRXnGfGzMkW\nNHX58bfqJrnLIaIUxoBGkunyDaKrexDFDiM0av5To4nhqouLkGnQ4rV/NuBIe5/c5RBRiuJPTZLM\nR4djm0gX2Y0yV0KUPBk6Da6bX4yoCPz61X0YDoblLomIUhADGkkiKorY+nE7NGoBs0sccpdDlFTF\nTjPKZzrg6Q3gj5sPQRS5DRQRnR4GNJLEp0e64eoZwoxJmTAZdHKXQ5R0l52bD6fVgO01XajaUseQ\nRkSnhQGNJPH2R60AgHNLuTgtTUwatQpLLi9FTpYeb+5uwfp36xnSiChhDGg07tw9Q9hX70WBLQNF\nTqvc5RDJxpihxc1XTofVrMOmD5rx0jv1CEe4/AYRnRoDGo27dz9ugwigbArXPiPKNGhRceV0ZJt0\neGNnM37y3C40dHB2JxGNjQGNxlUoHMH7eztg1KsxpzRP7nKIFMFs1OHOa2ZgbokNrW4/HvvDbjz/\nxkEcavYhGmW3JxEdTyN3AZRettd0YWAohIumW6DVqOUuh0gxMnQaLL6kGLOn2PDGziZs/aQdWz9p\nR6ZBi5nFVlgydcg26WDQaxCOiIhEoghHoghHxM/+jEYRiUQhikCWSYcskw7WTD0Kck1w2gxQq/g7\nN1G6YECjcdM3GMT6d+uhVQuYN4NLaxCdSHGeGd++bjaaXQM40OTFkfZ+7D7oOuvrajUqFOaacE6R\nBbOnWHFOkQUZOn6LJ0pVfPfSuKnaUouBoRC+OCcHVrNJ7nKIFEutVqEkPwsl+VkQRREDQyH4h8Pw\nD4UwOByARq2CWq2CRq2CRq2GRq2CSiVAPfK/CGAoEEafP4DegWG4ewbR3R9Ei2sAjZ39+MeuFqhV\nAkoKsjC72IpZxVaUFmZzRw+iFMKARuNib70XO2q6kG/To3z2JLnLIUoZgiDAbNTBbDyD9QLtmcc8\nDIWjaPf4caTdh2bXAOrbelHX2ovX/tkIvVaNmZMtKCuxoazEhjwbd/ggUjIGNDprQ4Ew/rj5IFQq\n4MsXFELN39KJZKHVqFCcZ0ZxnhkAMBwMo8U1gPo2H1pcfuyp92JPvRcAkJOVgQtnOTHFacKMIius\nZr2cpRPR5zCg0VkJhSP47ev74e0LoPwcCwrs2XKXREQjMnQaTJ9kwfRJFgBArz+II+09aGjvQat7\nCP/4oGn0uQ6rATOKLJgx2YIZRVbkZGfIVTYRARDENFna2u3ul7uECWc4GMbTL+/DgSYfJtsN+Obl\n00975qbFYkRPz6BEFZLUeP9SVzQqwh+K4tPaLrS4+tHuHUYw/NkiulazHlPzs1BSEBsrNyXPDIOe\nv9Mrhd1u5s+9FGa3m0/5HMnfbatXr8aePXsgCAIeeughzJ07d/RYdXU11qxZA7VajS996Uu47777\nTnkOKcPAUAhrX9qD+vY+TMs34voFpVxWgyiFqFQCipxmmPVqzEcssLl6htDU0YMWVz86fQF8eNiN\nDw+7AQACgIJcE4qcmcizGZFnMyI/xwSn1QCdlu99ovEmaUDbtWsXmpqaUFVVhfr6eqxcuRJVVVWj\nxx9//HE8++yzcDgcuP3223H11Veju7t7zHNIXoFQBFs+bMWmHU3wD4cxqygTi+eXcnYYUYpTqYTR\n4PWFMkAURfQPhtDmGUCbqw9dviG4egbR5vEfc54AwJaVAbslA1mm2GSHLJMOWUYtskYmP5gMGpgy\ntDBmaPi9gihBkga07du3Y9GiRQCA0tJS9PX1we/3w2QyoaWlBRaLBU6nEwCwcOFCbN++Hd3d3Sc9\nh+QRiUbR0N6PTxu82LqnHb0DQRh0alxWloP5c4sgCILcJRLROBMEYWQxXBtmFdsAxFrZ+gaD8PYO\nw+Xrh68/gJ6BIHoGQjjYPJzQdfU6NTIzPgtsJoMWpgwtzEYtzAYtzEYdMo3xx7GP9WyhowlI0oDm\n8XgwZ86c0cdWqxUejwcmkwkejwc2m230mM1mQ0tLC3w+30nPofEVCkfh6R1CIBRBIBiJ/RmKYjgY\nRv9gCC7fENw9Q2jq7MdgIAwA0GoElJ9jwcWzC2EynMGyAESUslQqAZZMPSyZepQWHjshKBKJYjAQ\nweBwCANDQfQPBjA4FMRgIDz6vSX2fwTDwQg6u0PHjHkbi06rGg1rRwc5s1GLTEMswGk1Kmg1aug0\nqpGPVVCrBAiCAEHAMX+qAEAAVMKxxyGKiIqx1sNoVEQUn30sikBUHPkzKkKEiEhURDgcRTga2/kh\nFBYRiUYRCkcRiYqxPyNRhCJRREZ2ggiH4ztDHL1LxGcfR6Ii1CohtgaeRgWtWgWtZuSxOvZ1adQq\nWLINCAyHoFUf+9zYx0Ls46M+p1YJQOw/4Ki/AwGxrx+I7U7BFk7lSOqIz7HmI5zsWJrMYVCkNX/5\nBAebe075PEumFtMKslBkN6G0KJerkxPRcdRqFcxGFcxGLZxIbI21SFTEcDCM4UAEg4EwBoYC8A8F\nMTgUxHAwguFQBEOBCIaDUQwFw2hzBxCO8GeCVCbZTfjJPV+QuwwaIelPWofDAY/HM/rY5XLBbreP\nHnO73aPHurq64HA4oNVqT3rOWBKZEUHH+uXShejqcsHjja2LJIqACEClUkOl1kCv00Ml+d5+IZgd\nWolfg6TD+5falHD/zqwlXhRFRCIRRKIRRCJhiKIIMRqBGP2sVW60dSz2IPa5kWMqlXDKBoD4YfGo\nT4ijn4+1pomiGPs+KQgQBFWsJU5QQaWK/a9WqaFSqcZ1KEj8a4+KUUSjUUQiEUCMxv4OxCgEfNa4\nEXvZk/89HM1iyebPUgWRNKAtWLAA69atw0033YSamho4nU4YjbHfrAoLC+H3+9He3g6Hw4F3330X\nTz75JLq7u096Do0/p9MBp5P7ZhIRESmJ5Oug/dd//Rd27twJtVqNVatWYf/+/TCbzVi0aBF2796N\nX/3qVwCAa665Bt/61rdOeM6MGTOkLJGIiIhIUdJmoVoiIiKidMHpGkREREQKw4BGREREpDAMaERE\nREQKk7IBLRKJYMWKFbj11ltRUVGBjz76CABw8OBBVFRU4NZbb8Wjjz4qc5V0Mjt37sSll16KrVu3\njn7ujjvuwJIlS3DHHXegsrIS+/fvl7FCGsuJ7h/fe6llw4YNuPzyy1FZWYnKykr85je/kbskStDq\n1atRUVGBW265Bfv27ZO7HErQzp07MX/+fFRWVuKOO+7AY489NubzU3bF0VdffRVGoxEvvvgi6urq\n8KMf/QgvvfQSfvazn+E///M/UVZWhuXLl+P999/HF7/4RbnLpaO0tLTgueeew4UXXnjcsSeeeAKl\npaUyVEWJOtn943sv9Vx77bV48MEH5S6DTsOp9rgmZSsvL8fatWsTem7KtqB97Wtfw4oVKwDEtonq\n7e1FKBRCa2srysrKAABXXnklqqur5SyTTsDhcOB//ud/kJmZedwxTipWvhPdv1AohLa2Nr73iCR2\nsj2uKTWczs+4lA1oarUaOl1sBernn38eN9xwA3w+HywWy+hzbDbbMbsVkDLo9fqTrqr91FNP4fbb\nb8fDDz+MYDCY5MooESe6fz6fD9nZn+3NyPdeati5cyfuvfde3HXXXThw4IDc5VACPr+PdXy/akoN\n9fX1uO+++3Dbbbed8pfYlOjifOmll7B+/XoIQmxrDkEQ8MADD2DBggV44YUXsH//fvz617+Gd2TL\nIlKOse7d5915552YMWMGioqK8Mgjj+CFF17AXXfdJUPVFHc694+U60T38brrrsMDDzyAhQsX4pNP\nPsGDDz6IjRs3yl0qnSb2OqSO4uJi3H///Vi8eDFaWlpQWVmJN998ExrNiaNYSgS0JUuWYMmSJcd9\n/qWXXsK7776LZ555Bmq1GjabDT6fb/R4fH9Pks/J7t2JxJvtAeCKK67AG2+8IVVZlKBE7x/fe8p2\nqvt4/vnnw+fzjYY3Uq6x9rgmZXM6nVi8eDEAoKioCLm5uejq6kJhYeEJn5+yXZwtLS3485//jHXr\n1kGrjW32q9FoMHXq1NEZnf/4xz84SFnhjv7t76677kJ/fz+AWNfL9OnT5SqLEhS/f3zvpZ7f/va3\n+Nvf/gYAOHz4MGw2G8NZCliwYAE2b94MANyvOsVs3LgRzz77LADA7XbD6/XC6XSe9Pkpu9XTmjVr\n8Pe//x35+fmjv/U9++yzaGpqwqpVqyCKIs477zz88Ic/lLtU+pytW7fit7/9LRoaGmCz2WC32/G7\n3/0OmzZtwv/+7//CZDLB4XDgZz/7GfR6vdzl0uec7P7V19fzvZdCurq68B//8R8QRRGRSAQ/+tGP\nMHfuXLnLogRwv+rU5Pf7sXz5cvT39yMcDuP+++8f8xfZlA1oREREROkqZbs4iYiIiNIVAxoRERGR\nwjCgERERESkMAxoRERGRwjCgERERESkMAxoRERGRwjCgEZEitLW1YeHCheN+3R/96Ee45pprUFlZ\niTvuuAMVFRV48803T+sa69atw9q1axN+/s6dO3Hrrbce93mPx4Pvf//7o3WtX7/+mM+5XC7s2LHj\ntGojovSUEls9EdHEINVK9t/+9rdx4403Aoit4P31r38d5eXlx2zwPt5O9LXk5ubiv//7v0/6uQ8+\n+AD19fW45JJLJKuLiFIDAxoRKd769evx5z//GQaDAbm5ufjpT38Kk8mE9evX4w9/+ANycnJw4YUX\norq6Gi+++OKY17Lb7XA6nWhtbcXbb7+Nd955B/39/fjWt76FOXPmYOXKlfD7/QiFQvj2t789ukds\nc3Mz/u3f/g0ulwvl5eVYsWIFhoaG8MMf/hC9vb3w+/24+uqrce+99wIAAoEAVqxYgcbGRmRmZuKp\np56Cz+fDrbfeiq1bt47W09bWhltvvRUvvvgi1qxZAyC2ddaf/vQnvPXWWzAYDAiFQrjiiivw97//\nHVlZWRL9LRORkjCgEZGidXR0YN26ddi0aRMMBgN+/vOf47nnnsOdd96JX/7yl9i0aRNsNhuWL1+e\nUAtcbW0tfD4fSktLcfjwYRw6dAh/+9vfoNFosGrVKpSXl+Puu+9Gd3c3vvrVr+LSSy8FADQ0NGD9\n+vWIRCK49tprceONNyIjIwOLFi3CV7/6VQSDQVx66aWjXZu1tbV45pln4HA48OCDD2LDhg24/PLL\nT1ijIAgoLCzEN77xDUQiEdx///3o6OjA5s2b8fWvfx3btm3DJZdcwnBGNIFwDBoRKVpNTQ3mzJkD\ng8EAAPjCF76Affv2obGxEYWFhbDZbACAq6+++qTX+N3vfjc6Bu3JJ5/E008/jYyMDADA7NmzodHE\nflfdu3fvaCCz2WzIy8tDQ0MDAODiiy+GSqWCVqvFnDlzUFtbi5ycHOzevRsVFRW45557EAwG0dvb\nCwCYOnUqHA4HAGDevHmoq6s7ra/75ptvxiuvvAIA2LRp02gXLRFNDGxBIyJF+3yLkyiKEAQB0Wj0\nmGMq1cl/37znnntOGnC0Wu1JX+vo1zj6+vEann/+eYRCIVRVVQHAMWPHjr5W/Pmn49xzz8XAwAAa\nGhpQV1fHcWlEEwxb0IhIMURRPO5zc+bMQU1NDQYHBwEA1dXVOP/88zF58mS0traiv78fAPDWW2+d\n9euff/752LZtGwCgq6sLHo8HJSUlAGIzM6PRKILBIGpqanDOOefA4/GgtLQUALBlyxYEAgEEg0EA\nwJEjR+B2uwEAH330Ec4555xTvr4gCAiFQqOPb7rpJqxcuRJXXXXVWX9tRJRa2IJGRIrh8/lQWVk5\n2uJ07rnn4gc/+AGWLl2Kb33rW9Dr9XA6nVi+fDkyMjLwr//6r7jllltQUFCAsrIytLe3n9XrP/DA\nA3jooYewdetWhEIh/PSnPx3tWp0+fTq+//3vo6WlBYsXL8bUqVNx4403YtmyZdi2bRu+/OUv44Yb\nbsAPfvAD/PCHP0RZWRnWrFmDpqYmmM1mfO1rX0N3d/eYr3/RRRdh2bJl0Ol0+N73vofrr78eq1ev\nPm7mJxGlP0E80a+sREQp4NVXX8UVV1yBrKwsPPfcc2hoaMCjjz4qd1njZtOmTdiyZQt+9atfyV0K\nESUZW9CIKGUNDg6isrISZrMZWq0Wq1evlrukcfO9730P3d3deOqpp+QuhYhkwBY0IiIiIoXhJAEi\nIiIihWFAIyIiIlIYBjQiIiIihWFAIyIiIlIYBjQiIiIihWFAIyIiIlKY/w/B6/gTIOrTNwAAAABJ\nRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f3ee4d12710>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import itertools\n",
    "import seaborn as sns\n",
    "\n",
    "sequences = reduce( lambda x, y: x+y, [[ seq for seq in it.product( 'ACGT', repeat=i ) ] for i in xrange( 1,6 )] )\n",
    "scores = map( model.log_probability, sequences )\n",
    "\n",
    "plt.figure( figsize=(10,5) )\n",
    "sns.kdeplot( numpy.array( scores ), shade=True )\n",
    "plt.ylabel('Density')\n",
    "plt.xlabel('Log Probability')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Training Hidden Markov Models"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are two main algorithms for training hidden Markov models-- Baum Welch (structured version of Expectation Maximization), and Viterbi training. Since we don't start off with labels on the data, these are both unsupervised training algorithms. In order to assign labels, Baum Welch uses EM to assign soft labels (weights in this case) to each point belonging to each state, and then using weighted MLE estimates to update the distributions. Viterbi assigns hard labels to each observation using the Viterbi algorithm, and then updates the distributions based on these hard labels.\n",
    "\n",
    "pomegranate is extremely well featured when it comes to regularization methods for training, supporting tied emissions and edges, edge and emission inertia, freezing nodes or edges, edge pseudocounts, and multithreaded training. Lets look at some examples of the following:\n",
    "\n",
    "### Tied Emissions\n",
    "\n",
    "Sometimes we want to say that multiple states model the same phenomena, but are simply at different points in the graph because we are utilizing complicated edge structure. An example is in the example of the global alignment HMM we saw. All insert states represent the same phenomena, which is nature randomly inserting a nucleotide, and this probability should be the same regardless of position. However, we can't simply have a single insert state, or we'd be allowed to transition from any match state to any other match state.\n",
    "\n",
    "You can tie emissions together simply by passing the same distribution object to multiple states. That's it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "d = NormalDistribution( 5, 2 )\n",
    "\n",
    "s1 = State( d, name=\"Tied1\" )\n",
    "s2 = State( d, name=\"Tied2\" )\n",
    "\n",
    "s3 = State( NormalDistribution( 5, 2 ), name=\"NotTied1\" )\n",
    "s4 = State( NormalDistribution( 5, 2 ), name=\"NotTied2\" )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You have now indicated that these two states are tied, and when training, the weights of all points going to s2 will be added to the weights of all points going to s1 when updating d. As a side note, this is implemented in a computationally efficient manner such that d will only be updated once, not twice (but giving the same result). s3 and s4 are not tied together, because while they have the same distribution, it is not the same python object.\n",
    "\n",
    "### Tied Edges\n",
    "\n",
    "Edges can be tied together for the same reason. If you have a modular structure to your HMM, perhaps you believe this repeating structure doesn't (or shouldn't) have a position specific edge structure. You can do this simply by adding a group when you add transitions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "model = HiddenMarkovModel()\n",
    "model.add_states( [s1, s2] )\n",
    "model.add_transition( model.start, s1, 0.5, group='a' )\n",
    "model.add_transition( model.start, s2, 0.5, group='b' )\n",
    "model.add_transition( s1, s2, 0.5, group='a' )\n",
    "model.add_transition( s2, s1, 0.5, group='b' )\n",
    "model.bake()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above model doesn't necessarily make sense, but it shows how simple it is to tie edges as well. You can go ahead and train normally from this point, without needing to change any code.\n",
    "\n",
    "### Inertia\n",
    "\n",
    "The next options are inertia on edges or on distributions. This simply means that you update your parameters as (previous_parameter * inertia) + (new_parameter * (1-inertia) ). It is a way to prevent your updates from overfitting immediately. You can specify this in the train function using either `edge_inertia` or `distribution_inertia`. These default to 0, with 1 being the maximum, meaning that you don't update based on new evidence, the same as freezing a distribution or the edges."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "Warning",
     "evalue": "Method is depricated. Use `fit` instead.",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mWarning\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-19-cec67a20ef33>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mmodel\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtrain\u001b[0m\u001b[1;33m(\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m5\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m2\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m3\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m4\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;36m5\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m7\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m2\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m3\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m5\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mdistribution_inertia\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m0.3\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0medge_inertia\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m0.25\u001b[0m \u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32mpomegranate/hmm.pyx\u001b[0m in \u001b[0;36mpomegranate.hmm.HiddenMarkovModel.train (pomegranate/hmm.c:26209)\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32mpomegranate/hmm.pyx\u001b[0m in \u001b[0;36mpomegranate.hmm.HiddenMarkovModel.train (pomegranate/hmm.c:25985)\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mWarning\u001b[0m: Method is depricated. Use `fit` instead."
     ]
    }
   ],
   "source": [
    "model.train( [[5, 2, 3, 4], [5, 7, 2, 3, 5]], distribution_inertia=0.3, edge_inertia=0.25 )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pseudocounts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Another way of regularizing your model is to add pseudocounts to your edges (which have non-zero probabilities). When updating your edges in the future, you add this pseudocount to the count of transitions across that edge in the future. This gives a more Bayesian estimate of the edge probability, and is useful if you have a large model and don't expect to cross most of the edges with your training data. An example might be a complicated profile HMM, where you don't expect to see deletes or inserts at all in your training data, but don't want to change from the default values.\n",
    "\n",
    "In pomegranate, pseudocounts default to the initial probabilities, so that if you don't see data, the edge values simply aren't updated. You can define both edge specific pseudocounts when you define the transition. When you train, you must define `use_pseudocount=True`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "s1 = State( NormalDistribution( 3, 1 ), name=\"s1\" )\n",
    "s2 = State( NormalDistribution( 6, 2 ), name=\"s2\" )\n",
    "\n",
    "model = HiddenMarkovModel()\n",
    "model.add_states( [s1, s2] )\n",
    "model.add_transition( model.start, s1, 0.5, pseudocount=4.2 )\n",
    "model.add_transition( model.start, s2, 0.5, pseudocount=1.3 )\n",
    "model.add_transition( s1, s2, 0.5, pseudocount=5.2 )\n",
    "model.add_transition( s2, s1, 0.5, pseudocount=0.9 )\n",
    "model.bake()\n",
    "model.train( [[5, 2, 3, 4], [5, 7, 2, 3, 5]], max_iterations=5, use_pseudocount=True )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The other way is to put a blanket pseudocount on all edges."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "s1 = State( NormalDistribution( 3, 1 ), name=\"s1\" )\n",
    "s2 = State( NormalDistribution( 6, 2 ), name=\"s2\" )\n",
    "\n",
    "model = HiddenMarkovModel()\n",
    "model.add_states( [s1, s2] )\n",
    "model.add_transition( model.start, s1, 0.5 )\n",
    "model.add_transition( model.start, s2, 0.5 )\n",
    "model.add_transition( s1, s2, 0.5 )\n",
    "model.add_transition( s2, s1, 0.5 )\n",
    "model.bake()\n",
    "model.train( [[5, 2, 3, 4], [5, 7, 2, 3, 5]], max_iterations=5, transition_pseudocount=20, use_pseudocount=True )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that there isn't as much of an improvement. This is part of regularization, though. We sacrifice fitting the data exactly in order for our model to generalize better to future data. The majority of the training improvement is likely coming from the emissions better fitting the data, though."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Multithreaded Training"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since pomegranate is implemented in cython, the majority of functions are written with the GIL released. A benefit of doing this is that we can use multithreading in order to make some computationally intensive tasks take less time. However, a downside is that python doesn't play nicely with multithreading, and so there are some cases where training using multithreading can make your model training take significantly longer. I investigate this in an early multithreading pull request <a href=\"https://github.com/jmschrei/pomegranate/pull/30\">here</a>. Things have improved since then, but the gist is that if you have a small model (less than 15 states), it may be detrimental, but the larger your model is, the more it scales towards getting a speed improvement exactly the number of threads you use. You can specify multithreading using the `n_jobs` keyword. All structures in pomegranate are thread safe, so you don't need to worry about race conditions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "s1 = State( NormalDistribution( 3, 1 ), name=\"s1\" )\n",
    "s2 = State( NormalDistribution( 6, 2 ), name=\"s2\" )\n",
    "\n",
    "model = HiddenMarkovModel()\n",
    "model.add_states( [s1, s2] )\n",
    "model.add_transition( model.start, s1, 0.5 )\n",
    "model.add_transition( model.start, s2, 0.5 )\n",
    "model.add_transition( s1, s2, 0.5 )\n",
    "model.add_transition( s2, s1, 0.5 )\n",
    "model.bake()\n",
    "model.train( [[5, 2, 3, 4, 7, 3, 6, 3, 5, 2, 4], [5, 7, 2, 3, 5, 1, 3, 5, 6, 2]], max_iterations=5 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "s1 = State( NormalDistribution( 3, 1 ), name=\"s1\" )\n",
    "s2 = State( NormalDistribution( 6, 2 ), name=\"s2\" )\n",
    "\n",
    "model = HiddenMarkovModel()\n",
    "model.add_states( [s1, s2] )\n",
    "model.add_transition( model.start, s1, 0.5 )\n",
    "model.add_transition( model.start, s2, 0.5 )\n",
    "model.add_transition( s1, s2, 0.5 )\n",
    "model.add_transition( s2, s1, 0.5 )\n",
    "model.bake()\n",
    "model.train( [[5, 2, 3, 4, 7, 3, 6, 3, 5, 2, 4], [5, 7, 2, 3, 5, 1, 3, 5, 6, 2]], max_iterations=5, n_jobs=4 )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Serialization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "General Mixture Models support serialization to JSONs using `to_json()` and `from_json( json )`. This is useful is you want to train a GMM on large amounts of data, taking a significant amount of time, and then use this model in the future without having to repeat this computationally intensive step (sounds familiar by now). Lets look at the original CG island model, since it's significantly smaller."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "seq = list('CGACTACTGACTACTCGCCGACGCGACTGCCGTCTATACTGCGCATACGGC')\n",
    "\n",
    "d1 = DiscreteDistribution({'A': 0.25, 'C': 0.25, 'G': 0.25, 'T': 0.25})\n",
    "d2 = DiscreteDistribution({'A': 0.10, 'C': 0.40, 'G': 0.40, 'T': 0.10})\n",
    "\n",
    "s1 = State( d1, name='background' )\n",
    "s2 = State( d2, name='CG island' )\n",
    "\n",
    "hmm = HiddenMarkovModel()\n",
    "hmm.add_states(s1, s2)\n",
    "hmm.add_transition( hmm.start, s1, 0.5 )\n",
    "hmm.add_transition( hmm.start, s2, 0.5 )\n",
    "hmm.add_transition( s1, s1, 0.5 )\n",
    "hmm.add_transition( s1, s2, 0.5 )\n",
    "hmm.add_transition( s2, s1, 0.5 )\n",
    "hmm.add_transition( s2, s2, 0.5 )\n",
    "hmm.bake()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "print hmm.to_json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "seq = list('CGACTACTGACTACTCGCCGACGCGACTGCCGTCTATACTGCGCATACGGC')\n",
    "print hmm.log_probability( seq )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "hmm_2 = HiddenMarkovModel.from_json( hmm.to_json() )\n",
    "print hmm_2.log_probability( seq )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
